File: AnteaterPathwayAPI/Models/Course.cs
using System;
using MongoDB.Bson.Serialization.Attributes;

namespace AnteaterPathwayAPI.Models
{
    [BsonIgnoreExtraElements]
    public class Course : MongoDbBaseModel
    {
        [BsonRequired][BsonElement("dept")]
        public string Department { get; set; }
        
        [BsonRequired][BsonElement("deptCode")]
        public string DepartmentCode { get; set; }
        
        [BsonRequired][BsonElement("num")]
        public string Number { get; set; }
        
        [BsonRequired][BsonElement("title")]
        public string Title { get; set; }
        
        [BsonIgnoreIfNull][BsonElement("unit")]
        public string Unit { get; set; }
        
        [BsonIgnoreIfNull][BsonElement("desc")]
        public string Description { get; set; }
        
        [BsonIgnoreIfNull][BsonElement("ge")]
        public string GeCategory { get; set; }
        
        [BsonIgnoreIfNull][BsonElement("restriction")]
        public string Restriction { get; set; }
        
        [BsonIgnoreIfNull][BsonElement("sameAs")]
        public string SameAs { get; set; }
        
        [BsonIgnoreIfNull][BsonElement("overlapsWith")]
        public string OverlapsWith { get; set; }
        
        [BsonIgnoreIfNull][BsonElement("concurrentWith")]
        public string ConcurrentWith { get; set; }
        [BsonIgnoreIfNull][BsonElement("gradingOption")]
        public string GradingOption { get; set; }
        
        [BsonIgnoreIfNull][BsonElement("repeatability")]
        public string Repeatability { get; set; }
        
        [BsonIgnoreIfNull][BsonElement("corequisite")]
        public string Corequisite { get; set; }
        
        [BsonIgnoreIfNull][BsonElement("preOrCorequisite")]
        public string PreOrCorequisite { get; set; }
    }
}

File: new-nextjs-client/src/models/course.ts
import { nanoid } from 'nanoid';

export interface CourseInfo {
  deptCode: string;
  num: string;
  title: string;
  unit: number | null;
  isVariableUnit: boolean;
  isWorkloadCredit: boolean;
  minUnit: number | null;
  maxUnit: number | null;
  geCode: string | null;
  offered: string[];
  prerequisite: string | null;
  corequisite: string | null;
  prereqOrCoreq: string | null;
  sameAs: string | null;
  concurrentWith: string | null;
  overlapsWith: string | null;
  gradingOption: string | null;
  repeatability: string | null;
  restriction: string | null;
}

export class Course {
  private _id: string;
  public isCustomCreated: boolean = false;
  public color = 1;
  public courseInfo: CourseInfo;

  constructor(courseInfo: CourseInfo, isCustomCreated: boolean) {
    this._id = nanoid();
    this.courseInfo = courseInfo;
    this.isCustomCreated = isCustomCreated;
  }

  public get id() {
    return `${this.courseInfo.deptCode}-${this.courseInfo.num}-${this.color}-${this._id}`;
  }
}
